### ISSUE TITLE
Gateways: exclude the meta item "access public read" and "access public write" from the sync between gateway and central manager

### ISSUE THREAD

**Issue by DonWillems** [openremote/openremote#1435]
If a gateway is connected, and attributes should be available publicly there locally, those config items should not be synced with the central instance. Centrally you would not want the attributes to be public.

**Comment by richturner** [comment]
GatewayConnector should strip out unwanted meta items during synchronisation (when edge gateway tells the central instance which assets & attributes it has). This should be implemented in:

`GatewayConnector.saveAssetLocally()`

Ideally MetaItemDescriptor should indicate whether it should be synchronised from edge gateway to central manager, but a simple hard coded list may be sufficient for now:

* AGENT_LINK
* ATTRIBUTE_LINKS
* ACCESS_PUBLIC_READ
* ACCESS_PUBLIC_WRITE
* ACCESS_RESTRICTED_READ
* ACCESS_RESTRICTED_WRITE

How do we handle items like:

* STORE_DATA_POINTS
* DATA_POINTS_MAX_AGE_DAYS
* FORECAST etc..

**Comment by pankalog** [comment]
> How do we handle items like:
> 
> * STORE_DATA_POINTS
> * DATA_POINTS_MAX_AGE_DAYS
> * FORECAST etc..

@richturner Those should probably be added to the data configuration in the gateway configuration? I can see situations where we would like to store datapoints locally for 3 months and keep it in the main cloud instance indefinitely or something

**Comment by DonWillems** [comment]
The user should be able to overwrite the meta items in this list, and only those in this list, on the central manager.
This means an attribute could have ACCESS_RESTRICTED_READ on the central manager and ACCESS_PUBLIC_READ on the edge gateway. The edge will not sync that meta item with the central instance, so it will not be overwritten, and the user can change it to have restricted read.

@pankalog @richturner Does that make sense?
This will still be quite some manual work if new gateways are connected to the system regularly, having to set the ACCESS_RESTRICTED_READ for every attribute that needs it. So not very elegant for now.

**Comment by richturner** [comment]
> > How do we handle items like:
> > 
> > * STORE_DATA_POINTS
> > * DATA_POINTS_MAX_AGE_DAYS
> > * FORECAST etc..
> 
> [@richturner](https://github.com/richturner) Those should probably be added to the data configuration in the gateway configuration? I can see situations where we would like to store datapoints locally for 3 months and keep it in the main cloud instance indefinitely or something

Nice thinking about utilising the Gateway Asset to store the rules but I'm struggling to see how we could define the items there and then keep that synchronised across the child assets

**Comment by Hackerberg43** [comment]
Just for some additional clarity of our specific use case:

- We have assets with ACCESS_RESTRICTED_READ/WRITE, and STORE DATA POINTS ,  on most attributes. 
     - The assets are created in the gateway via auto-provisioning. 
     - These meta-items are hardcoded in the custom asset class.
     - The enabling of meta items could be (partly) moved to the auto-provisioning asset template if necessary.

- The meta-item ACCESS_RESTRICTED is required to be present in the master-manager, because the asset owners are restricted users there.

- We would like to add the ACCESS_PUBLIC_READ item to some attributes, but only to the asset in the gateway, because we'd like to create some public dashboards in the local network, but not in the cloud.

I am not aware of the asset synchronization method, but if the master manager instance would start with duplicating the asset based on the custom asset class itself, I would expect that hardcoded meta items will be enabled also in the master manager, via this first creation process, even when there is no meta-item synchronization anymore.

So if all of this is true, then choosing whether to add the meta-item to the asset class or auto-provisioning template, would decide if the meta item configuration will be identical or gateway specific.

Of course this would not apply to later added custom attributes and manually created assets.

**Comment by richturner** [comment]
Hey thanks for the info @Hackerberg43 

The implementation that is being done should match your requirements; it may not be a long term solution and might be reworked but we wanted to get some solution in for now.

Asset creation within the manager backend performs JSR-380 validation using the `AssetValidator` so any user/code submitting assets for merging to the DB is responsible for ensuring that it meets the requirements of the `AssetTypeDescriptor` and `AttributeDescriptor`(s).

There's no validation on meta items but the manager UI will set them based on the `AttributeDescriptor`(s) of the selected asset type.

For auto provisioning the provisioning config asset template must conform to the asset type otherwise it will fail.

Hope this clarifies things a little.

**Comment by richturner** [comment]
An update:

Now looking to make attribute and meta item filtering on a per asset type basis in the `GatewayConnection` so it will be configured on the edge gateway and filtering will be applied before assets are set to the central instance.

GatewayConnection additional field:
```protected Map<String, GatewayAssetSyncRule> assetSyncRules;```

Example:
```
{
   "*": { <-- Applies to all assets
      "excludeAttributes": ["notes"],
      "excludeMetaItems": {
          "*": ["accessPublicRead", "accessPublicWrite", "accessRestrictedRead"],
          "location": [] <-- Don't exclude any meta items for location attribute
      }
   },
   "CityAsset": {
      "excludeAttributes": ["city"]
   }
}
```

* If a definition exists for a specific asset type then that should be used instead of any `*` wildcard definition
* If a definition exists for a specific attribute within `excludeMetaItems` then that should be used instead of any `*` wildcard definition

**Comment by richturner** [comment]
#1628 is a proposal to allow force re-sync of descendant assets which could be useful when altering sync rules on the edge gateway.

**Comment by Hackerberg43** [comment]
You are taking this to a whole new level.

I've read your forceRefreshOnNextConnect issue, but dont't fully understand it. Is it to bypass the sync rules ?
The term "refresh" suggests that a bypassed metaitem sync will also happen during initial gateway creation/connection?

**Comment by richturner** [comment]
> You are taking this to a whole new level.
> 
> I've read your forceRefreshOnNextConnect issue, but dont't fully understand it. Is it to bypass the sync rules ? The term "refresh" suggests that a bypassed metaitem sync will also happen during initial gateway creation/connection?

* Currently assets are created/overwritten in the central instance during gateway sync at connection time (one way forced sync)
* To allow different meta items to be added to gateway descendant attributes in the central manager we cannot then do the same one way forced sync during connection time (as this would overwrite any meta items added at the central instance)

The above means some strategy is needed to handle the merging of edge gateway asset and the central instance representation of this asset.

Currently we were looking at only syncing meta items from edge -> central when the asset and/or attribute doesn't already exist so that means on existing deployments when the sync rules are created/changed on the edge gateway nothing would happen to the meta items on the central instance so you would need to manually go through every attribute and remove meta items.

This is why the `forceRefreshOnNextConnect` issue was created so when set the next time the edge gateway connects the meta items on the central instance would be replaced with whatever the edge gateway sends (after sync rules applied) so you can effectively reset the meta items.

Dealing with the merging is a challenge, having the edge gateway as the source of truth simplifies this so rather than just stripping out meta items before sending to the central instance we could look to apply some sort of mapping (remove X,Y,Z meta items and add A,B,C meta items), any thoughts on this would be appreciated.

**Comment by Hackerberg43** [comment]
Right, got it now; enforce the removal of meta items in central manager selected not to be synced in the gateway.

The single source of truth makes sense, also more clear as a user I think. So disabling the possibility to alter attribute meta items from the central manager. **However, that would be a lot more effort while not building on the intermediate solution created now?**
Might have to discuss about the hours, the goal for us was "just a little tweak" to be able to make public dashboards in the local network, but this is getting more architectural.

To add on the thoughts though for single point of truth:

An intermediate solution could be:
-Start with making it only possbile from the json editor
-Make the mapping on a per attribute basis

Options:
-Sync : default, current behavior, but not able to alter in central manager
-Add : add the metaitem in central manager, even when it doesnt exist in the gateway
-Remove: Remove the metaitem in central manager, even when it doesnt exist in the gateway

Then later, the UI could be extended to allow easily group selecting metaitems by:
-Central: apply to all metaitems for all attributes of all assets;
-asset: apply to all metaitems for all attributes of selected asset;
-attribute: apply to metaitems of selected asset.


Something like this:

![Image](https://github.com/user-attachments/assets/898a9a6c-98fb-4610-b140-5edbfd8eee96)

**Comment by richturner** [comment]
So we will aim to do the following for now:

Attribute and/or Meta item configuration will be made available in the Manager interconnect UI of the edge gateway as a JSON editor field for now (UI to be added in #1654) and will provide meta item mapping functionality (remove and add meta items on a per attribute basis) with the edge gateway being the source of truth and so synching is one directional (from edge gateway to central instance).

In the central instance gateway assets cannot be modified (same as existing functionality).

GatewayConnection additional field:
```protected Map<String, GatewayAssetSyncRule> assetSyncRules;```

Example:
```
{
   "*": { <-- Applies to all assets
      "excludeAttributes": ["notes"],
      "excludeMetaItems": {
          "*": ["accessPublicRead"],
          "location": [] <-- Don't exclude any meta items for location attribute
      },
      "addMetaItems": {
          "*": { <-- MetaMap
             "accessRestrictedRead": true
          },
      }
   },
   "CityAsset": {
      "excludeAttributes": ["city"]
   }
}
```

* If a definition exists for a specific asset type or attribute then that should be used instead of any `*` wildcard definition
* If a definition exists for a specific attribute within `excludeMetaItems` then that should be used instead of any `*` wildcard definition

**Comment by richturner** [comment]
Docs added: https://github.com/openremote/documentation/pull/58
